Fixes #11836 -- Missing django.forms.widgets.MultiWidget hidden counterpart

smithdc1 opened the pull request:
[Ticket](https://code.djangoproject.com/ticket/11836)

Currently `MultipleHiddenInput` gives each subwidget the same name. This pull request appends an index reference and therefore gives consistency to other widgets. 

I've added an attribute so that the new behaviour can be turned off, to enable backwards compatibility. I've currently defaulted to `True` as I think this is the right thing to do (it improves consistency) and it also shows the impact on the entire test suite. 

I think the key 'design' decisions are:

- Do we need the flag to enable the behaviour to be change at all
- Do we put this on a deprecation path, to remove the flag in future
- Do we default to True/False

hramezani commented on tests/forms_tests/widget_tests/test_multiplehiddeninput.py:
```diff
@@ -68,8 +69,29 @@ def test_render_increment_id(self):
+            ),
+        )
+
+    def test_unique_names_render_attrs(self):
```
Do we really need this test? seems it doesn't relate to this change.

smithdc1 replied:
Hi @hramezani 
I'm trying to show passing tests for both `unique_names = True` and `unique_name=False`. Would it be preferable to reduce it to just a single test for `unique_names=False`. Appreciate it's a bit repetitive.

hramezani replied:
Hi @smithdc1, I meant ‍`test_unique_names_render_attrs‍`. Did your changes affect on attrs rendering?

smithdc1 replied:
Hi @hramezani no change to attrs rendering. I added a test to prove that this is the case. Do you suggest that this test is not required?

hramezani replied:
I think we don't need it. but lets @felixxm decide about it.
Thanks for the patch :+1:

felixxm commented on django/forms/widgets.py:
```diff
@@ -353,9 +357,13 @@ def get_context(self, name, value, attrs):
+            if self.unique_names:
+                widget_name = '%s_%s' % (name, index)
+            else:
+                widget_name = name
```
IMO we don't need a new parameter, the previous behavior was incorrect. We should use different `name`'s for all widgets (like `MultiWidget`).

\cc @carltongibson

carltongibson replied:
Yes, I agree: let's just fix the broken behaviour, adding the release note as per [comment](https://github.com/django/django/pull/12567/files#r394830854). 

There's no (non-array, i.e. `name[]`) use-case where multiple inputs with the same `name` is the right behaviour.

carltongibson replied:
Having said that, I'm immediately not even sure that we need the release note: `MultipleHiddenInput` is just broken no? (I guess it doesn't hurt...)

felixxm commented on docs/releases/3.1.txt:
```diff
@@ -213,6 +213,11 @@ Forms
+
+* :class:`django.forms.MultipleHiddenInput` now accepts a
+  ``unique_names`` attribute. When set to True this results in index being
+  appended to the field name. Defaults to True.
```
We should add a note to the `Backwards incompatible changes in 3.1`.

smithdc1 commented:
Hi @felixxm thanks for your feedback. 

I've updated this based on your feedback (I've also rebased to remove merge conflict). 

The one element I was unsure on was the release notes. It seemed a bit repetitive to mention such a small change twice. So I moved it to the incompatible changes section as you suggested.

felixxm commented:
@smithdc1 Thanks for updates :+1: However, I have doubts about this ticket after checking [docs](https://docs.djangoproject.com/en/3.0/ref/forms/widgets/#multiplehiddeninput) and tests. `MultipleHiddenInput` is not a "hidden" version of `MultiWidget` it's a widget _"that handles multiple hidden widgets for fields that have **a list of values**"_, so the current behavior with using the same **name** is correct.

I think we should close this ticket as wontfix (\cc @carltongibson).

smithdc1 commented:
Hi @felixxm 

One other [comment](https://code.djangoproject.com/ticket/11836#comment:5) on the ticket suggested getting values didn't work. However, looks ok on master to me. Example with this PR also provided for comparison.

This PR
```
w = MultipleHiddenInput()
w.render("example", ["value1", "value2"])
'<input type="hidden" name="example_0" value="value1"><input type="hidden" name="example_1" value="value2">'
data = {"example" : ["value1", "value2"]}
w.value_from_datadict(data, {}, "example")
['value1', 'value2']
```

Master

```
w = MultipleHiddenInput()
w.render("example", ["value1", "value2"])
'<input type="hidden" name="example" value="value1"><input type="hidden" name="example" value="value2">'
data = {"example" : ["value1", "value2"]}
w.value_from_datadict(data, {}, "example")
['value1', 'value2']
```

felixxm commented:
On PR we have an example of using `MultipleHiddenInput` as a `hidden_widget` for `MultiValueField` which cannot work because it's not the same use case as `MultiWidget` (multiple fields vs. multiple values), see fb60a6ff0acc84417ced061e83bb170fff351b59.

felixxm commented:
@smithdc1 Many thanks for refreshing this old ticket :+1: Let's close it.
